posthog/hogql/expr_parser.py

```diff
@@ -0,0 +1,187 @@
+import ast
+import re
+from dataclasses import dataclass, field
+from typing import List, Optional
+
+from clickhouse_driver.util.escape import escape_param
+
+from posthog.models.property.util import get_property_string_expr
+
+EVENT_FIELDS = ["id", "uuid", "event", "timestamp", "distinct_id"]
+PERSON_FIELDS = ["id", "created_at", "properties"]
+CLICKHOUSE_FUNCTIONS = ["concat", "coalesce"]
+HOGQL_AGGREGATIONS = ["avg", "sum", "total"]
+
+
+@dataclass
+class ExprParserContext:
+    property_list: List[List[str]] = field(default_factory=list)
+    encountered_nodes: List[ast.AST] = field(default_factory=list)
+    is_aggregation: bool = False
+
+
+def translate_hql(hql: str, context: Optional[ExprParserContext] = None) -> str:
+    """Translate a HogQL expression into a Clickhouse expression."""
+    try:
+        # Until we swap out the AST parser, we're limited to Python's dialect.
+        # This means "properties.$bla" fails. The following is a hack to get around that fofr now.
+        hql = re.sub(r"properties\.(\$[\$a-zA-Z0-9_\-]+)", r"properties['\1']", hql)
+        node = ast.parse(hql)
+    except SyntaxError as err:
+        raise ValueError(f"SyntaxError: {err.msg}")
+    if not context:
+        context = ExprParserContext()
+    return translate_ast(node, [], context)
+
+
+def translate_ast(node: ast.AST, stack: List[ast.AST], context: ExprParserContext) -> str:
+    """Translate a parsed HogQL expression in the shape of a Python AST into a Clickhouse expression."""
+    response = ""
+    stack.append(node)
+    context.encountered_nodes.append(node)
+    if type(node) == ast.Module:
```

Comment:
AFAIK `isinstance()` is more "pythonic" and is able to deal with inheritance, though probably does not make a difference here

Reply:
I tried it earlier, but had something with `mypy`, so switched to this. Now that I refactored it back to `isinstance`, it seems good... 🤷